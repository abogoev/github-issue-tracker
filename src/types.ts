import { NativeStackScreenProps } from "@react-navigation/native-stack";

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export enum CurrentRoute {
  Home = "Home",
  Details = "Details",
}

export type StackParams = {
  [CurrentRoute.Home]: undefined;
  [CurrentRoute.Details]: { owner: string; repo: string; number: number };
};

export type HomeNavigationProps = NativeStackScreenProps<
  StackParams,
  typeof CurrentRoute.Home
>;

export type DetailsNavigationProps = NativeStackScreenProps<
  StackParams,
  typeof CurrentRoute.Details
>;

export type Issue = {
  number: number;
  title: string;
  user: {
    avatar_url: string;
  };
  body: string;
};
