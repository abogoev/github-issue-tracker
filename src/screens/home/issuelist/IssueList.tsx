import { useNavigation } from "@react-navigation/native";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import React, { useCallback, VFC } from "react";
import { FlatList, ListRenderItem, StyleSheet, ViewStyle } from "react-native";
import IssueItem from "../../../core/issueitem/IssueItem";
import { RootStackParamList } from "../../../router/Router";
import { IssueIntercept } from "../../../types";
import { DETAILS_SCREEN } from "../../../utils/constants";

interface Props {
  style?: ViewStyle;
  issues: IssueIntercept[];
}

const IssueList: VFC<Props> = ({ issues, style, ...props }) => {
  const navigation =
    useNavigation<NativeStackNavigationProp<RootStackParamList>>();

  const renderItem: ListRenderItem<IssueIntercept> = useCallback(
    ({ item }) => (
      <IssueItem
        title={item.title}
        number={item.number}
        createdAt={item.created_at}
        closedAt={item.closed_at && new Date(item.closed_at)}
        user={item.user.login}
        onPress={() =>
          navigation.navigate(DETAILS_SCREEN, {
            owner: item.owner,
            repo: item.repo,
            number: item.number,
          })
        }
        style={style}
      />
    ),
    []
  );

  return (
    <FlatList
      data={issues}
      renderItem={renderItem}
      keyExtractor={(item, index) => index.toString()}
      {...props}
    />
  );
};

export default IssueList;

const styles = StyleSheet.create({});
