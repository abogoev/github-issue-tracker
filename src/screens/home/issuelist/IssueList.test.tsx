import React from "react";
import "react-native";
import IssueList from "./IssueList";
import { create } from "react-test-renderer";
import { IssueIntercept } from "../../../types";

jest.mock("@react-navigation/core");

describe("IssueList", () => {
  it("should render correctly", () => {
    const issues: IssueIntercept[] = [
      {
        title: "Title",
        number: 1,
        created_at: "2012-04-23T18:25:43.511Z",
        closed_at: null,
        owner: "Owner",
        repo: "Repo",
        body: "asd",
        user: { avatar_url: "", login: "username" },
      },
    ];
    const component = create(<IssueList data={issues} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
