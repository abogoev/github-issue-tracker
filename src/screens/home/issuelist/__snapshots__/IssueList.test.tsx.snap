// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueList should render correctly 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "body": "asd",
        "closed_at": null,
        "created_at": "2012-04-23T18:25:43.511Z",
        "number": 1,
        "owner": "Owner",
        "repo": "Repo",
        "title": "Title",
        "user": Object {
          "avatar_url": "",
          "login": "username",
        },
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#fff",
            "borderColor": "rgb(208, 215, 222)",
            "borderWidth": 1,
            "minHeight": 62,
            "opacity": 1,
            "padding": 8,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <svg
            data-file-name="SvgOpenIssue"
            style={
              Object {
                "marginHorizontal": 8,
              }
            }
          />
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#24292f",
                  "fontFamily": "Helvetica",
                  "fontSize": 16,
                  "fontWeight": "600",
                  "lineHeight": 24,
                }
              }
            >
              Title
            </Text>
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "rgb(87, 96, 106)",
                  "fontFamily": "Helvetica",
                  "fontSize": 12,
                  "lineHeight": 18,
                  "marginTop": 4,
                }
              }
            >
              #1 opened on 23/04/2012, 21:25:43 by username
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
